File: test/services/py/client_roundtrip.py
"""In-process round-trip test for the generated PetStore RPC client.

Exercises the generated `PetStoreClient` (UEBA + JSON method variants) against an
in-process transport that delegates to the generated server-side wiring invokers
(`invoke_ueba_PetStore` / `invoke_json_PetStore`). No sockets: the transport
callback feeds the client-encoded bytes/str straight into the wiring and returns
the wiring's encoded response for the client to decode.

Run directly: `python3 client_roundtrip.py`
"""
from Generated.baboon_codecs import BaboonCodecContext
from Generated.baboon_service_wiring import BaboonMethodId
from Generated.petstore.api.PetStore import PetStore
from Generated.petstore.api.PetStore_Client import PetStoreClient
from Generated.petstore.api import PetStore_Wiring

invoke_json_PetStore = getattr(PetStore_Wiring, "invoke_json_PetStore", None)
invoke_ueba_PetStore = getattr(PetStore_Wiring, "invoke_ueba_PetStore", None)

from Generated.petstore.api.Pet import Pet
from Generated.petstore.api.PetStatus import PetStatus
from Generated.petstore.api.petstore.addpet.In import In as AddPetIn
from Generated.petstore.api.petstore.addpet.Out import Out as AddPetOut
from Generated.petstore.api.petstore.getpet.In import In as GetPetIn
from Generated.petstore.api.petstore.getpet.Out import Out as GetPetOut
from Generated.petstore.api.petstore.listpets.In import In as ListPetsIn
from Generated.petstore.api.petstore.listpets.Out import Out as ListPetsOut
from Generated.petstore.api.petstore.deletepet.In import In as DeletePetIn
from Generated.petstore.api.petstore.deletepet.Out import Out as DeletePetOut


class PetStoreImpl(PetStore):
    def __init__(self) -> None:
        self._pets: dict[int, Pet] = {}
        self._next_id: int = 1

    def addPet(self, arg: AddPetIn) -> AddPetOut:
        pet_id = self._next_id
        self._next_id += 1
        pet = Pet(id=pet_id, name=arg.name, status=arg.status, tag=arg.tag)
        self._pets[pet_id] = pet
        return AddPetOut(pet=pet)

    def getPet(self, arg: GetPetIn) -> GetPetOut:
        pet = self._pets.get(arg.id)
        if pet is None:
            raise KeyError(f"Pet not found: id={arg.id}")
        return GetPetOut(pet=pet)

    def listPets(self, arg: ListPetsIn) -> ListPetsOut:
        return ListPetsOut(pets=sorted(self._pets.values(), key=lambda p: p.id))

    def deletePet(self, arg: DeletePetIn) -> DeletePetOut:
        existed = arg.id in self._pets
        if existed:
            del self._pets[arg.id]
        return DeletePetOut(deleted=existed)


def _make_client(impl: PetStoreImpl, ctx: BaboonCodecContext) -> PetStoreClient:
    # The generated client constructor takes only the transports for the codecs
    # that are active (UEBA first, then JSON, then ctx). Build the positional
    # argument list to match whatever was generated.
    args: list = []
    if invoke_ueba_PetStore is not None:
        def transport_ueba(service: str, method: str, data: bytes) -> bytes:
            assert service == "PetStore", f"unexpected service {service}"
            return invoke_ueba_PetStore(BaboonMethodId(service, method), data, impl, ctx)
        args.append(transport_ueba)
    if invoke_json_PetStore is not None:
        def transport_json(service: str, method: str, data: str) -> str:
            assert service == "PetStore", f"unexpected service {service}"
            return invoke_json_PetStore(BaboonMethodId(service, method), data, impl, ctx)
        args.append(transport_json)
    args.append(ctx)
    return PetStoreClient(*args)


def _run(use_json: bool) -> None:
    impl = PetStoreImpl()
    ctx = BaboonCodecContext.compact()
    client = _make_client(impl, ctx)

    if use_json:
        add = client.addPet_json
        get = client.getPet_json
        lst = client.listPets_json
        dele = client.deletePet_json
        label = "JSON"
    else:
        add = client.addPet
        get = client.getPet
        lst = client.listPets
        dele = client.deletePet
        label = "UEBA"

    buddy = add(AddPetIn(name="Buddy", status=PetStatus.Available, tag="dog"))
    assert buddy.pet.name == "Buddy", f"[{label}] expected Buddy, got {buddy.pet.name}"
    buddy_id = buddy.pet.id

    whiskers = add(AddPetIn(name="Whiskers", status=PetStatus.Pending, tag="cat"))
    whiskers_id = whiskers.pet.id

    listed = lst(ListPetsIn())
    assert len(listed.pets) == 2, f"[{label}] expected 2 pets, got {len(listed.pets)}"

    got = get(GetPetIn(id=buddy_id))
    assert got.pet.name == "Buddy", f"[{label}] expected Buddy, got {got.pet.name}"
    assert got.pet.status == PetStatus.Available, f"[{label}] expected Available, got {got.pet.status}"
    assert got.pet.tag == "dog", f"[{label}] expected tag dog, got {got.pet.tag}"

    deleted = dele(DeletePetIn(id=whiskers_id))
    assert deleted.deleted is True, f"[{label}] expected deleted=True, got {deleted.deleted}"

    listed2 = lst(ListPetsIn())
    assert len(listed2.pets) == 1, f"[{label}] expected 1 pet, got {len(listed2.pets)}"
    assert listed2.pets[0].name == "Buddy", f"[{label}] expected remaining Buddy, got {listed2.pets[0].name}"

    print(f"{label} client round-trip OK")


def run_client_roundtrip() -> None:
    ran = False
    if invoke_ueba_PetStore is not None:
        _run(use_json=False)
        ran = True
    if invoke_json_PetStore is not None:
        _run(use_json=True)
        ran = True
    assert ran, "no active codecs: neither JSON nor UEBA wiring was generated"
    print("OK")


if __name__ == "__main__":
    run_client_roundtrip()

File: test/services/jv-client/src/main/java/example/ClientRoundTrip.java
package example;

import baboon.runtime.shared.BaboonClientTransport;
import baboon.runtime.shared.BaboonCodecContext;
import baboon.runtime.shared.BaboonMethodId;
import petstore.api.PetStatus;
import petstore.api.PetStoreClient;
import petstore.api.PetStoreWiring;

import java.util.Optional;

/**
 * In-process round-trip for the GENERATED {@link PetStoreClient}:
 *
 *   client encode -> in-process transport -> server PetStoreWiring.invoke* -> decode
 *
 * Exercises both the JSON-suffixed client methods (transportJson +
 * invokeJson) and the UEBA default-named client methods (transportUeba +
 * invokeUeba). The transport callbacks are pure in-process dispatch into
 * the server wiring against a shared {@link PetStoreImpl}; no sockets.
 */
public final class ClientRoundTrip {
    private ClientRoundTrip() {}

    private static void check(boolean condition, String message) {
        if (!condition) {
            throw new AssertionError(message);
        }
    }

    public static void main(String[] args) throws Exception {
        BaboonCodecContext ctx = BaboonCodecContext.Default;
        PetStoreImpl impl = new PetStoreImpl();

        BaboonClientTransport.JsonSync jsonTransport =
            (service, method, data) -> PetStoreWiring.invokeJson(new BaboonMethodId(service, method), data, impl, ctx);
        BaboonClientTransport.UebaSync uebaTransport =
            (service, method, data) -> PetStoreWiring.invokeUeba(new BaboonMethodId(service, method), data, impl, ctx);

        // The generated sync client takes both transports + ctx.
        PetStoreClient client = new PetStoreClient(uebaTransport, jsonTransport, ctx);

        // ---- JSON path (the *Json client methods) ----
        impl.reset();
        runScenario("JSON", client, true);

        // ---- UEBA path (the bare-named client methods) ----
        impl.reset();
        runScenario("UEBA", client, false);

        System.out.println("OK");
    }

    private static void runScenario(String label, PetStoreClient client, boolean json) throws Exception {
        var addBuddyIn = new petstore.api.petstore.addpet.In("Buddy", PetStatus.Available, Optional.of("dog"));
        var addBuddyOut = json ? client.addPetJson(addBuddyIn) : client.addPet(addBuddyIn);
        long buddyId = addBuddyOut.pet().id();
        check(addBuddyOut.pet().name().equals("Buddy"), label + ": expected name Buddy, got " + addBuddyOut.pet().name());

        var addWhiskersIn = new petstore.api.petstore.addpet.In("Whiskers", PetStatus.Pending, Optional.of("cat"));
        var addWhiskersOut = json ? client.addPetJson(addWhiskersIn) : client.addPet(addWhiskersIn);
        long whiskersId = addWhiskersOut.pet().id();
        check(addWhiskersOut.pet().name().equals("Whiskers"), label + ": expected name Whiskers, got " + addWhiskersOut.pet().name());

        var listIn = new petstore.api.petstore.listpets.In();
        var listOut = json ? client.listPetsJson(listIn) : client.listPets(listIn);
        check(listOut.pets().size() == 2, label + ": expected 2 pets, got " + listOut.pets().size());

        var getBuddyIn = new petstore.api.petstore.getpet.In(buddyId);
        var getBuddyOut = json ? client.getPetJson(getBuddyIn) : client.getPet(getBuddyIn);
        check(getBuddyOut.pet().name().equals("Buddy"), label + ": expected Buddy, got " + getBuddyOut.pet().name());
        check(getBuddyOut.pet().status() == PetStatus.Available, label + ": expected Available, got " + getBuddyOut.pet().status());
        check(getBuddyOut.pet().tag().isPresent() && getBuddyOut.pet().tag().get().equals("dog"),
            label + ": expected tag dog, got " + getBuddyOut.pet().tag());

        var deleteIn = new petstore.api.petstore.deletepet.In(whiskersId);
        var deleteOut = json ? client.deletePetJson(deleteIn) : client.deletePet(deleteIn);
        check(deleteOut.deleted(), label + ": expected deleted=true, got false");

        var list2Out = json ? client.listPetsJson(listIn) : client.listPets(listIn);
        check(list2Out.pets().size() == 1, label + ": expected 1 pet, got " + list2Out.pets().size());
        check(list2Out.pets().get(0).name().equals("Buddy"), label + ": expected remaining pet Buddy, got " + list2Out.pets().get(0).name());

        System.out.println(label + " round-trip OK");
    }
}
